pkg/util/debouncer/debouncer.go

```diff
@@ -0,0 +1,298 @@
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var (
+	ErrBufferFull = errors.New("debouncer buffer full")
+)
+
+type KeyFunc[T any] func(T) string
+type ProcessFunc[T any] func(context.Context, T) error
+
+type metrics struct {
+	itemsAddedCounter           prometheus.Counter
+	itemsDroppedCounter         prometheus.Counter
+	itemsProcessedCounter       prometheus.Counter
+	processingErrorsCounter     prometheus.Counter
+	processingDurationHistogram prometheus.Histogram
+}
+
+func newMetrics(reg prometheus.Registerer, name string) *metrics {
+	return &metrics{
+		itemsAddedCounter: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "debouncer_items_added_total",
+			Help: "Total number of items added to the debouncer",
+			ConstLabels: prometheus.Labels{
+				"name": name,
+			},
+		}),
+		itemsDroppedCounter: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "debouncer_items_dropped_total",
+			Help: "Total number of items dropped due to a full buffer",
+			ConstLabels: prometheus.Labels{
+				"name": name,
+			},
+		}),
+		itemsProcessedCounter: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "debouncer_items_processed_total",
+			Help: "Total number of items processed by the debouncer",
+			ConstLabels: prometheus.Labels{
+				"name": name,
+			},
+		}),
+		processingErrorsCounter: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "debouncer_processing_errors_total",
+			Help: "Total number of errors during processing",
+			ConstLabels: prometheus.Labels{
+				"name": name,
+			},
+		}),
+		processingDurationHistogram: promauto.With(reg).NewHistogram(prometheus.HistogramOpts{
+			Name:                            name + "_debouncer_processing_duration_seconds",
+			Help:                            "Time taken to process items",
+			Buckets:                         instrument.DefBuckets,
+			NativeHistogramBucketFactor:     1.1,
+			NativeHistogramMaxBucketNumber:  160,
+			NativeHistogramMinResetDuration: time.Hour,
+			ConstLabels: prometheus.Labels{
+				"name": name,
+			},
+		}),
+	}
+}
+
+type DebouncerOpts[T any] struct {
+	BufferSize   int
+	KeyFunc      KeyFunc[T]
+	MinWait      time.Duration
+	MaxWait      time.Duration
+	Reg          prometheus.Registerer
+	Name         string
+	ErrorHandler func(T, error)
+}
+
+type Debouncer[T any] struct {
+	opts   DebouncerOpts[T]
+	buffer chan T
+
+	// mutex protecting the debouncers map.
+	debouncersMu sync.Mutex
+	debouncers   map[string]*keyDebouncer[T]
+
+	wg           sync.WaitGroup
+	ctx          context.Context
+	cancel       context.CancelFunc
+	errorHandler func(T, error)
+	metrics      *metrics
+}
+
+// NewDebouncer creates a new debouncer for processing events with unique keys.
+//
+// A debouncer helps optimize expensive operations by:
+// 1. Grouping identical events that occur in rapid succession
+// 2. Processing each unique key only once after waiting periods expire
+//
+// Parameters:
```

Comment:
Hopefully I didn't duplicate this comment. 

I think it would be helpful to include a note about the Name parameter. This is mainly to highlight that the metrics rely on it for labelling and therefore prevent initialisation of a debouncer group without one. I suppose we'll have to do the same for the registerer since it's also used to initialise metrics.

Reply:
I would suggest to move options documentation to fields comment in  `DebouncerOpts` and document all of them (as Arati mentions, `Name` is missing for example).

Reply:
Done 👍